### PR TITLE
18) added DriverDaoJdbcImpl, CarDaoJdbcImpl

### PR DESCRIPTION
**Pull request by AntonPanchishko**

### CODE REVIEW

File: src/main/java/taxi/dao/jdbc/CarDaoJdbcImpl.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+            statement.setLong(1, car.getManufacturer().getId());
+            statement.setString(2, car.getModel());
+            statement.executeUpdate();
+            ResultSet resultSet = statement.getGeneratedKeys();
+            if (resultSet.next()) {
+                car.setId(resultSet.getObject(1, Long.class));
+            }
+            return car;
+        } catch (SQLException ex) {
+            throw new DataProcessingException("Couldn't add car into DB ", ex);
```
Comment (vicire) [inline src/main/java/taxi/dao/jdbc/CarDaoJdbcImpl.java]:
```suggestion
            throw new DataProcessingException("Couldn't add car " + car + " into DB ", ex);
```

File: src/main/java/taxi/dao/jdbc/CarDaoJdbcImpl.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            statement.setLong(1, id);
+            ResultSet resultSet = statement.executeQuery();
+            Car car = null;
+            if (resultSet.next()) {
+                car = createCarFromDb(resultSet);
+                car.setDrivers(getDriversForCar(id, connection));
+            }
+            return Optional.of(car);
```
Comment (vicire) [inline src/main/java/taxi/dao/jdbc/CarDaoJdbcImpl.java]: What happened if a car will be empty?

File: src/main/java/taxi/dao/jdbc/CarDaoJdbcImpl.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+            return cars;
+        } catch (SQLException ex) {
+            throw new DataProcessingException("Can't get list car by driver id = "
+                    + driverId, ex);
+        }
+    }
+
+    private Car createCarFromDb(ResultSet resultSet) throws SQLException {
+        Long id = resultSet.getObject("car_id", Long.class);
+        Manufacturer manufacturer = ManufacturerDaoJdbcImpl.createManufacturerFromDb(resultSet);
```
Comment (vicire) [inline src/main/java/taxi/dao/jdbc/CarDaoJdbcImpl.java]: that`s optional but I think that using ManufacturerDaoJdbcImpl is not a good idea. Better to make a separate method here.

File: src/main/java/taxi/dao/jdbc/CarDaoJdbcImpl.java [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        try (Connection connection = ConnectionUtil.getConnection();
+                   PreparedStatement statementForCars = connection.prepareStatement(queryForCars)) {
+            statementForCars.setLong(1, car.getManufacturer().getId());
+            statementForCars.setString(2, car.getModel());
+            statementForCars.setLong(3, car.getId());
+            statementForCars.executeUpdate();
+        } catch (SQLException ex) {
+            throw new DataProcessingException("Can't update car " + car.getModel(), ex);
+        }
+        deleteObjectInCarDriverById(car.getId());
```
Comment (vicire) [inline src/main/java/taxi/dao/jdbc/CarDaoJdbcImpl.java:101]: You open one more connection here and under

File: src/main/java/taxi/dao/jdbc/CarDaoJdbcImpl.java [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+            while (resultSet.next()) {
+                drivers.add(DriverDaoJdbcImpl.createDriverFromDb(resultSet));
+            }
+            return drivers;
+        } catch (SQLException e) {
+            throw new DataProcessingException("We can't get drivers for this car", e);
+        }
+    }
+
+    private boolean insertNewDateInCarDriver(Long driverId, Long carId) {
```
Comment (vicire) [inline src/main/java/taxi/dao/jdbc/CarDaoJdbcImpl.java]: you can make this method void

File: src/main/java/taxi/dao/jdbc/CarDaoJdbcImpl.java [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+                        connection.prepareStatement(queryForInsert)) {
+            statementForInsert.setLong(1, driverId);
+            statementForInsert.setLong(2, carId);
+            return statementForInsert.executeUpdate() > 0;
+        } catch (SQLException ex) {
+            throw new DataProcessingException("Cant insert into car_driver db", ex);
+        }
+    }
+
+    private boolean deleteObjectInCarDriverById(Long id) {
```
Comment (vicire) [inline src/main/java/taxi/dao/jdbc/CarDaoJdbcImpl.java]: Same here

File: src/main/java/taxi/dao/jdbc/DriverDaoJdbcImpl.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+            statement.setString(1, driver.getName());
+            statement.setString(2, driver.getLicenceNumber());
+            statement.executeUpdate();
+            ResultSet resultSet = statement.getGeneratedKeys();
+            if (resultSet.next()) {
+                driver.setId(resultSet.getObject(1, Long.class));
+            }
+            return driver;
+        } catch (SQLException ex) {
+            throw new DataProcessingException("Couldn't add driver into DB ", ex);
```
Comment (vicire) [inline src/main/java/taxi/dao/jdbc/DriverDaoJdbcImpl.java]: Same as in create method in carDao

File: src/main/java/taxi/dao/jdbc/ManufacturerDaoJdbcImpl.java [hunk 8]
```diff
@@ -103,7 +103,7 @@ public boolean delete(Long id) {
         }
     }
 
-    private Manufacturer createManufacturerFromDb(ResultSet resultSet) throws SQLException {
+    public static Manufacturer createManufacturerFromDb(ResultSet resultSet) throws SQLException {
```
Comment (vicire) [inline src/main/java/taxi/dao/jdbc/ManufacturerDaoJdbcImpl.java]: not a good practice, coz now all user can use this method in other classes

**Review by vicire — CHANGES_REQUESTED** [review]

**Review by vicire — APPROVED** [review]

File: src/main/java/taxi/dao/jdbc/CarDaoJdbcImpl.java [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+            statementForCars.setLong(3, car.getId());
+            statementForCars.executeUpdate();
+        } catch (SQLException ex) {
+            throw new DataProcessingException("Can't update car " + car.getModel(), ex);
+        }
+        deleteObjectInCarDriverById(car.getId());
+        for (int i = 0; i < car.getDrivers().size(); i++) {
+            insertNewDateInCarDriver(car.getDrivers().get(i).getId(), car.getId());
+        }
+        return car;
```
Comment (dmytro-hryhoruk) [inline src/main/java/taxi/dao/jdbc/CarDaoJdbcImpl.java]: let's use one connection for each insert and not create the new one on each iteration

File: src/main/java/taxi/dao/jdbc/CarDaoJdbcImpl.java [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+        }
+    }
+
+    @Override
+    public List<Car> getAllByDriver(Long driverId) {
+        String query = "SELECT cd.car_id "
+                + "FROM car_driver cd "
+                + "INNER JOIN driver d "
+                + "ON cd.driver_id = d.driver_id "
+                + "WHERE d.driver_id = ? AND deleted = false ";
```
Comment (dmytro-hryhoruk) [inline src/main/java/taxi/dao/jdbc/CarDaoJdbcImpl.java]: this deleted = false works for both car and driver? Let's better specify the table name and write this condition twice for car and for dirver

**Review by dmytro-hryhoruk — CHANGES_REQUESTED** [review]

**Review by dmytro-hryhoruk — APPROVED** [review]

**Pull request merged**